### PR TITLE
Fee market v3

### LINKED ISSUE #227: Fee market issue
https://github.com/darwinia-network/darwinia-messages-substrate/issues/168

### PR DESCRIPTION
**Pull request by hujw77**
# Merge after #224
- fee market v3
- fee market v3 unit test

Fix #227 
https://github.com/darwinia-network/darwinia-messages-substrate/pull/169#issuecomment-1197660979

### CODE REVIEW

**Comment by HackFisher** [comment]
@hujw77 Please help fix the confilicts.

File: contracts/bridge/src/fee-market/FeeMarket.sol [hunk 1]
```diff
@@ -385,70 +377,131 @@ contract FeeMarket is IFeeMarket {
...
+        uint256 vault_reward
+    ) {
+        Order memory order = orderOf[key];
+        require(orderOf[key].assignedTime > 0, "!exist");
+        // Get the message fee from the top N relayers
         uint256 message_fee = getOrderFee(key);
+        // Get slot index and slot price
+        (uint256 slot, uint256 slot_price) = _get_slot_price(key, message_fee);
+        // Message surplus = Message Fee - Slot price
+        uint256 message_surplus = message_fee - slot_price;
```
Comment (HackFisher) [inline contracts/bridge/src/fee-market/FeeMarket.sol:396]: Do we need to check overflow here?

File: contracts/bridge/src/fee-market/FeeMarket.sol [hunk 2]
```diff
@@ -385,70 +377,131 @@ contract FeeMarket is IFeeMarket {
...
-            address assign_relayer = assignedRelayers[key][slot].assignedRelayer;
-            // the message delivery in the `slot` assign_relayer
-            // [slot, slot+1)
-            if (slot * relayTime <= diff_time && diff_time < (slot + 1 ) * relayTime) {
-                uint256 base_fee = assignedRelayers[key][slot].makerFee;
-                (delivery_reward, confirm_reward, vault_reward) = _distribute_ontime(message_fee, base_fee, assign_relayer, delivery_relayer, confirm_relayer);
+        number = order.assignedRelayersNumber;
+        collateral = order.collateral;
+        // Diff_time = settle_time - assign_time
+        diff_time = block.timestamp - order.assignedTime;
```
Comment (HackFisher) [inline contracts/bridge/src/fee-market/FeeMarket.sol:422]: overflow check?

File: contracts/bridge/src/fee-market/FeeMarket.sol (hunk 1 above)
Comment (hujw77) [reply contracts/bridge/src/fee-market/FeeMarket.sol:396]: https://github.com/darwinia-network/darwinia-messages-sol/blob/a6ae7140e80d82598ef5e801e6a3946687607e9b/contracts/bridge/src/fee-market/FeeMarket.sol#L248

message_fee is the last top N assigned relayer, and also the max fee of the three assigned relayers.

File: contracts/bridge/src/fee-market/FeeMarket.sol (hunk 2 above)
Comment (hujw77) [reply contracts/bridge/src/fee-market/FeeMarket.sol:422]: `settle_time` > `assign_time` in the same order.

**Review by HackFisher — APPROVED** [review]

**Pull request merged**